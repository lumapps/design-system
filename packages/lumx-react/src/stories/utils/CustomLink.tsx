/* eslint-disable jsx-a11y/anchor-has-content,jsx-a11y/anchor-is-valid */
import React from 'react';

/**
 * Example custom link that can be used in linkAs props.
 */
export const CustomLink: any = ({ ...props }) => <a href="#" data-custom-link {...props} />;
