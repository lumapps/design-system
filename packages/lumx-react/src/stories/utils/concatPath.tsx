export type PathPart = string | number;
export type Path = PathPart | Array<PathPart>;

/**
 * Concat flatten object path
 *
 * @example concatPath('foo', 'bar') // => 'foo.bar'
 * @example concatPath(['foo', 0]) // => 'foo[0]'
 * @example concatPath('foo', 0, ['bar']) // => 'foo[0].bar'
 */
export const concatPath = (...prefix: Path[]) => {
    const [first, ...rest] = prefix.flat();
    return rest.reduce<string>((acc, part) => {
        if (typeof part === 'number') return `${acc}[${part}]`;
        return `${acc}.${part}`;
    }, String(first));
};
