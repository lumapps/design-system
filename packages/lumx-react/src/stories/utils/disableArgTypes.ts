export function disableArgTypes(argTypes: string[]) {
    return Object.fromEntries(argTypes.map((argType) => [argType, { table: { disable: true } }]));
}
