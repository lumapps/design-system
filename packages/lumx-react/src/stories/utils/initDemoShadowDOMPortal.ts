/** Init a demo shadow DOM to use as portal container */
export const initDemoShadowDOMPortal = () => {
    const div = document.createElement('div');
    document.body.appendChild(div);
    const container = div.attachShadow({ mode: 'open' });
    const style = document.createElement('style');
    style.innerText = `button { color: red; }`;
    container.appendChild(style);
    return { container, teardown: () => div.remove() };
};
