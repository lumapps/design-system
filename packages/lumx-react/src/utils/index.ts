/**
 * WARNING: All modules exported here are exposed to NPM in '@lumx/react/utils'.
 */

export { ClickAwayProvider } from './ClickAwayProvider';
export { Portal, type PortalProps, type PortalInit, PortalProvider, type PortalProviderProps } from './Portal';
