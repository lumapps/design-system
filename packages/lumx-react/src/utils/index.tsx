export * from '@lumx/core/js/utils';

export * from './getRootClassName';

export * from './htmlDecode';

export * from './partitionMulti';

export * from './type';
