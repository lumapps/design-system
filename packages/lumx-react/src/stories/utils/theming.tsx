import React from 'react';
import { Predicate } from '@lumx/react/utils/type';
import { FlexBox } from '@lumx/react';

type Property = string | { value: string };
interface PropertyTree {
    [key: string]: PropertyTree | Property;
}
type Entry = [key: string, value: string];
type Variables = Record<string, string>;

function asValue(prop?: PropertyTree | Property): string | undefined {
    if (typeof prop === 'string') {
        return prop;
    }
    if (typeof prop === 'object' && 'value' in prop) {
        return prop.value as string;
    }
    return undefined;
}

/** Convert property tree into a list of args entries ([key, value]) */
function* toArgEntries(props: PropertyTree | Property, parent = ''): Iterable<Entry> {
    const value = asValue(props);
    if (value) {
        yield [parent, value];
    } else {
        for (const [key, prop] of Object.entries(props)) {
            const newParent = parent ? `${parent}-${key}` : key;
            yield* toArgEntries(prop, newParent);
        }
    }
}
function* parseVariableEntries(cssString: string, strict = false) {
    for (const line of cssString.split(/(?<=;)/)) {
        if (!line.trim()) continue;
        const match = line.match(/^\s*(--lumx[^:]+)\s*:\s*([^;]+);/m);
        if (!match) {
            if (strict) throw new Error(`Could not parse \`${line}\``);
            else continue;
        }
        const [, variableName, variableValue] = match;
        yield [variableName, variableValue];
    }
}

function VariableEditor({
    variable: [key, value],
    setVariable,
    defaultStyle,
}: {
    variable: Entry;
    setVariable: (newVariable: Entry) => void;
    defaultStyle: Variables;
}) {
    const onChange = React.useCallback<React.ChangeEventHandler<HTMLInputElement>>(
        (evt) => {
            setVariable([evt.target.name, evt.target.value]);
        },
        [setVariable],
    );

    return (
        <FlexBox as="label" orientation="horizontal" hAlign="center" style={{ whiteSpace: 'nowrap' }}>
            <span style={{ flexShrink: 0 }}>{key}</span>:
            <input
                name={key}
                onChange={onChange}
                style={{ border: 0, padding: 8, fieldSizing: 'content', display: 'inline' } as any}
                value={defaultStyle[key] === value ? '' : value}
                placeholder={value}
            />
            ;
        </FlexBox>
    );
}

function VariablesEditor({
    style,
    defaultStyle,
    setStyle,
}: {
    style: Variables;
    defaultStyle: Variables;
    setStyle: (fn: (newStyle: Variables) => Variables) => void;
}) {
    const setVariable = React.useCallback(
        ([key, value]: Entry) => {
            setStyle((prevStyle) => ({ ...prevStyle, [key]: value }));
        },
        [setStyle],
    );
    return (
        <details style={{ fontFamily: 'monospace' }} className="lumx-spacing-margin-vertical-huge">
            <summary className="lumx-typography-subtitle2">Edit variables</summary>
            {Object.entries(style).map((variable) => (
                <VariableEditor
                    key={variable[0]}
                    defaultStyle={defaultStyle}
                    variable={variable}
                    setVariable={setVariable}
                />
            ))}
        </details>
    );
}

const FILTER_PROPS = ['theme', 'emphasis', 'size'];

export function withTheming<P extends PropertyTree>({
    properties,
    values,
}: {
    properties?: P;
    values?: string | Partial<P>;
}) {
    const initialStyle = values
        ? Object.fromEntries(
              typeof values === 'string'
                  ? parseVariableEntries(values)
                  : toArgEntries(values as PropertyTree, '--lumx'),
          )
        : [];
    const defaultStyle = Object.fromEntries(properties ? Array.from(toArgEntries(properties, '--lumx')) : []);

    const varNames = new Set(Object.keys(initialStyle).concat(Object.keys(defaultStyle)));

    function filteredStyles(args: any) {
        // Create predicate
        const filterArgs = Object.entries(args)
            .map(([argName, argValue]) => {
                if (typeof argValue === 'string' && FILTER_PROPS.includes(argName)) {
                    return (varName: string) =>
                        // Has the given arg with value
                        varName.match(new RegExp(`${argName}-${argValue}`)) ||
                        (argName === 'emphasis' && argValue === 'medium' && varName.match(/emphasis-selected/)) ||
                        // Or does not have the arg at all
                        !varName.match(new RegExp(`${argName}-\\w+`));
                }
                return false;
            })
            .filter(Boolean) as Array<Predicate<string>>;
        const styles = Object.fromEntries(
            Array.from(varNames.values())
                .map((varName) => {
                    if (!filterArgs.every((filter) => filter(varName))) return null as any;
                    return [varName, initialStyle[varName] || defaultStyle[varName]];
                })
                .filter(Boolean),
        );
        return () => styles;
    }

    return (Story: any, ctx: any) => {
        const [style, setStyle] = React.useState<Variables>(filteredStyles(ctx.args));

        return (
            <>
                <Story args={{ ...ctx.args, style: { ...ctx.args.style, ...style } }} />
                <VariablesEditor style={style} defaultStyle={defaultStyle} setStyle={setStyle} />
            </>
        );
    };
}
