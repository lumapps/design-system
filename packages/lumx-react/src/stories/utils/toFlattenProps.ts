import castArray from 'lodash/castArray';
import { concatPath } from './concatPath';

type Props = Record<string, any>;
type OneOrMoreProps = Props | Array<Props>;

/**
 * Build a flat props object from the given nested props
 *
 * @example toFlattenProps({ foo: { bar: 4 } }) // => { 'foo.bar': 4 }
 * @example toFlattenProps({ foo: [{ bar: 4 }, { bar: 5 }] }) // => { 'foo[0].bar': 4, 'foo[0].bar': 5 }
 */
export function toFlattenProps(props: { [prefix: string]: OneOrMoreProps }): Props {
    const out: Props = {};
    for (const [prefix, oneOrMoreProps] of Object.entries(props)) {
        const propsArray = castArray(oneOrMoreProps);

        for (let i = 0; i < propsArray.length; i += 1) {
            const path = Array.isArray(oneOrMoreProps) ? concatPath(prefix, i) : prefix;
            const subProps = propsArray[i];

            for (const [key, value] of Object.entries(subProps)) {
                out[concatPath(path, key)] = value;
            }
        }
    }
    return out;
}
