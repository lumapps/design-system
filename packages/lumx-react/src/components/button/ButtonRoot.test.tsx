import React from 'react';

import { mount } from 'enzyme';
import 'jest-enzyme';

import { commonTestsSuite, Wrapper } from '@lumx/react/testing/utils';

import { ColorPalette, Emphasis, Size, Theme } from '@lumx/react';
import {
    BUTTON_CLASSNAME,
    BUTTON_WRAPPER_CLASSNAME,
    ButtonRoot,
    ButtonRootProps,
} from '@lumx/react/components/button/ButtonRoot';
import { getBasicClass } from '@lumx/react/utils';

type SetupProps = Partial<ButtonRootProps>;

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 */
const setup = (propsOverride: SetupProps = {}) => {
    const props: any = { ...propsOverride };
    const wrapper: Wrapper = mount(<ButtonRoot {...props} />);

    return {
        button: wrapper.find(`.${BUTTON_CLASSNAME}`),
        buttonWrapper: wrapper.find(`.${BUTTON_WRAPPER_CLASSNAME}`),
        props,
        wrapper,
    };
};

describe(`<${ButtonRoot.displayName}>`, () => {
    // 1. Test render via snapshot (default states of component).
    describe('Snapshots and structure', () => {
        it('should render button with label', () => {
            const props: Partial<ButtonRootProps> = {
                children: 'Label',
            };
            const { button, buttonWrapper, wrapper } = setup(props);
            expect(wrapper).toMatchSnapshot();

            expect(buttonWrapper).not.toExist();

            expect(button).toExist();
            expect(button.type()).toEqual('button');

            expect(button.contains(props.children)).toBe(true);
        });

        it('should render button with wrapper and label', () => {
            const props: Partial<ButtonRootProps> = {
                children: 'Label',
                hasBackground: true,
            };
            const { button, buttonWrapper, wrapper } = setup(props);
            expect(wrapper).toMatchSnapshot();

            expect(buttonWrapper).toExist();

            expect(button).toExist();
            expect(button.type()).toEqual('button');

            expect(button.contains(props.children)).toBe(true);
        });

        it('should render anchor button with label', () => {
            const props: Partial<ButtonRootProps> = {
                children: 'Label',
                href: 'example.com',
                target: '_blank',
            };
            const { button, buttonWrapper, wrapper } = setup(props);
            expect(wrapper).toMatchSnapshot();

            expect(buttonWrapper).not.toExist();

            expect(button).toExist();
            expect(button.type()).toEqual('a');
            expect(button.contains(props.children)).toBe(true);

            const actualProps = button.props() as Partial<ButtonRootProps>;
            expect(actualProps.href).toEqual(props.href);
            expect(actualProps.target).toEqual(props.target);
        });

        it('should render anchor button with wrapper and label', () => {
            const props: Partial<ButtonRootProps> = {
                children: 'Label',
                hasBackground: true,
                href: 'example.com',
                target: '_blank',
            };
            const { button, buttonWrapper, wrapper } = setup(props);
            expect(wrapper).toMatchSnapshot();

            expect(buttonWrapper).toExist();

            expect(button).toExist();
            expect(button.type()).toEqual('a');
            expect(button.contains(props.children)).toBe(true);

            const actualProps = button.props() as Partial<ButtonRootProps>;
            expect(actualProps.href).toEqual(props.href);
            expect(actualProps.target).toEqual(props.target);
        });

        it('should render custom link', () => {
            const CustomLink: React.FC = ({ children, ...props }) =>
                React.createElement('a', { ...props, style: { color: 'red' } }, children);
            const { wrapper } = setup({ href: 'foo', linkAs: CustomLink });
            expect(wrapper).toMatchSnapshot();
        });
    });

    // 2. Test defaultProps value and important props custom values.
    describe('Props', () => {
        it('should use default color', () => {
            const props: Partial<ButtonRootProps> = {};
            const { button, wrapper } = setup(props);
            expect(wrapper).toMatchSnapshot();

            expect(button).toExist();
            expect(button).toHaveClassName(
                getBasicClass({ prefix: BUTTON_CLASSNAME, type: 'color', value: ColorPalette.dark }),
            );
        });

        it('should not have default color in low or medium emphasis', () => {
            const props: Partial<ButtonRootProps> = {
                emphasis: Emphasis.low,
            };
            const { button, wrapper } = setup(props);
            expect(wrapper).toMatchSnapshot();

            expect(button).not.toHaveClassName(
                getBasicClass({ prefix: BUTTON_CLASSNAME, type: 'color', value: ColorPalette.primary }),
            );
        });

        it('should use given props', () => {
            // Props used for the button.
            const buttonProps: Partial<ButtonRootProps> = {
                color: ColorPalette.red,
                emphasis: Emphasis.high,
                size: Size.s,
                theme: Theme.dark,
                variant: 'icon',
            };
            // Props used for the button wrapper.
            const buttonWrapperProps: Partial<ButtonRootProps> = {
                variant: buttonProps.variant,
            };
            const { button, buttonWrapper, wrapper } = setup({
                ...buttonProps,
                hasBackground: true,
            });
            expect(wrapper).toMatchSnapshot();

            // The button wrapper classes.
            for (const [type, value] of Object.entries(buttonProps)) {
                expect(button).toHaveClassName(getBasicClass({ prefix: BUTTON_CLASSNAME, type, value }));
            }

            // The button wrapper classes.
            for (const [type, value] of Object.entries(buttonWrapperProps)) {
                expect(buttonWrapper).toHaveClassName(getBasicClass({ prefix: BUTTON_WRAPPER_CLASSNAME, type, value }));
            }
        });
    });

    // 3. Test events.
    describe('Events', () => {
        // Nothing to do here.
    });

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Conditions', () => {
        // Tested in step 1.
    });

    // 5. Test state.
    describe('State', () => {
        // Nothing to do here.
    });

    // Common tests suite.
    commonTestsSuite(setup, { className: 'button', prop: 'button' }, { className: BUTTON_CLASSNAME });
});
