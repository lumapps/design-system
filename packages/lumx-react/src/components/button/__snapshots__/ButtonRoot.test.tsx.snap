// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonRoot> Props should not have default color in low or medium emphasis 1`] = `
<ButtonRoot
  emphasis="low"
>
  <button
    className="lumx-button lumx-button--color-dark lumx-button--emphasis-low"
    type="button"
  />
</ButtonRoot>
`;

exports[`<ButtonRoot> Props should use default color 1`] = `
<ButtonRoot>
  <button
    className="lumx-button lumx-button--color-dark"
    type="button"
  />
</ButtonRoot>
`;

exports[`<ButtonRoot> Props should use given props 1`] = `
<ButtonRoot
  color="red"
  emphasis="high"
  hasBackground={true}
  size="s"
  theme="dark"
  variant="icon"
>
  <div
    className="lumx-button-wrapper lumx-button-wrapper--variant-icon"
  >
    <ButtonRoot
      color="red"
      emphasis="high"
      hasBackground={false}
      size="s"
      theme="dark"
      variant="icon"
    >
      <button
        className="lumx-button lumx-button--color-red lumx-button--emphasis-high lumx-button--size-s lumx-button--theme-dark lumx-button--variant-icon"
        type="button"
      />
    </ButtonRoot>
  </div>
</ButtonRoot>
`;

exports[`<ButtonRoot> Snapshots and structure should render anchor button with label 1`] = `
<ButtonRoot
  href="example.com"
  target="_blank"
>
  <a
    className="lumx-button lumx-button--color-dark"
    href="example.com"
    target="_blank"
  >
    Label
  </a>
</ButtonRoot>
`;

exports[`<ButtonRoot> Snapshots and structure should render anchor button with wrapper and label 1`] = `
<ButtonRoot
  hasBackground={true}
  href="example.com"
  target="_blank"
>
  <div
    className="lumx-button-wrapper"
  >
    <ButtonRoot
      color="dark"
      hasBackground={false}
      href="example.com"
      target="_blank"
    >
      <a
        className="lumx-button lumx-button--color-dark"
        href="example.com"
        target="_blank"
      >
        Label
      </a>
    </ButtonRoot>
  </div>
</ButtonRoot>
`;

exports[`<ButtonRoot> Snapshots and structure should render button with label 1`] = `
<ButtonRoot>
  <button
    className="lumx-button lumx-button--color-dark"
    type="button"
  >
    Label
  </button>
</ButtonRoot>
`;

exports[`<ButtonRoot> Snapshots and structure should render button with wrapper and label 1`] = `
<ButtonRoot
  hasBackground={true}
>
  <div
    className="lumx-button-wrapper"
  >
    <ButtonRoot
      color="dark"
      hasBackground={false}
    >
      <button
        className="lumx-button lumx-button--color-dark"
        type="button"
      >
        Label
      </button>
    </ButtonRoot>
  </div>
</ButtonRoot>
`;

exports[`<ButtonRoot> Snapshots and structure should render custom link 1`] = `
<ButtonRoot
  href="foo"
  linkAs={[Function]}
>
  <CustomLink
    className="lumx-button lumx-button--color-dark"
    href="foo"
  >
    <a
      className="lumx-button lumx-button--color-dark"
      href="foo"
      style={
        Object {
          "color": "red",
        }
      }
    />
  </CustomLink>
</ButtonRoot>
`;
